tools/lcpp.patch

```diff
@@ -43,7 +46,7 @@ index 5ab65ea9..44fe3fe6 100644
      // get hparams kv
      ml.get_key(LLM_KV_VOCAB_SIZE, hparams.n_vocab, false) || ml.get_arr_n(LLM_KV_TOKENIZER_LIST, hparams.n_vocab);
  
-@@ -15854,26 +15867,40 @@ static ggml_type llama_tensor_get_type(quantize_state_internal & qs, ggml_type n
+@@ -15854,26 +15870,40 @@ static ggml_type llama_tensor_get_type(quantize_state_internal & qs, ggml_type n
```

Comment:
We'll have to adjust the logic below this line for SD1/SDXL/etc so the in/out/embed/etc keys get kept in the right precision. I can double check this once we merge so this is more of a note to myself though you can add it like this as a start:

```cpp
    if ( // KEEP IN FP32
            (name == tn(LLM_TENSOR_OUTPUT, "weight")) ||
            (!qs.has_output && name == tn(LLM_TENSOR_TOKEN_EMBD, "weight")) ||
            // flux
            (name.find("img_in.") != std::string::npos) ||
            (name.find("time_in.in_layer.") != std::string::npos) ||
            (name.find("vector_in.in_layer.") != std::string::npos) ||
            (name.find("guidance_in.in_layer.") != std::string::npos) ||
            (name.find("final_layer.linear.") != std::string::npos) ||
            // SD
            (name.find("conv_in.") != std::string::npos) || 
            (name.find("conv_out.") != std::string::npos) ||
            (name == "input_blocks.0.0.weight") ||
            (name == "out.2.weight")
        ) {
```

```cpp
    } else if ( // KEEP IN FP16
            (name == "token_embd.weight") ||
            // flux
            (name.find("txt_in.") != std::string::npos) ||
            (name.find("time_in.") != std::string::npos) ||
            (name.find("vector_in.") != std::string::npos) ||
            (name.find("guidance_in.") != std::string::npos) ||
            (name.find("final_layer.") != std::string::npos) ||
            // SD
            (name.find("time_embed.") != std::string::npos) ||
            (name.find("time_embedding.") != std::string::npos) ||
            (name.find("label_emb.") != std::string::npos) ||
            (name.find("add_embedding.") != std::string::npos)
        ) {
```